drivers/SmartThings/zwave-garage-door-opener/src/ecolink-zw-gdo/init.lua

```diff
@@ -0,0 +1,300 @@
+                          ECOLINK_GARAGE_DOOR_FINGERPRINTS.productId)
+end
+
+local function component_to_endpoint(device, component_id)
+  if (CONTACTSENSOR_ENDPOINT_NAME == component_id)  then
+    --contactSensor is 2
+    return CONTACTSENSOR_ENDPOINT_NUMBER
+  end
+  -- main endpoint is garage door
+  return GDO_ENDPOINT_NUMBER
+end
+
+local function endpoint_to_component(device, ep)
+  if ( CONTACTSENSOR_ENDPOINT_NUMBER == ep ) then
+    return CONTACTSENSOR_ENDPOINT_NAME
+  end
+  return GDO_ENDPOINT_NAME
+end
+
+--- Handle Device Instantiated Event
+---
+--- @param driver st.zwave.Driver
+--- @param device st.zwave.Device
+local function device_instantiated(driver, device)
+  log.info_with({hub_logs=true}, "device init")
+  device:set_component_to_endpoint_fn(component_to_endpoint)
+  device:set_endpoint_to_component_fn(endpoint_to_component)
+  device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY, function(d)
+    device:send(BarrierOperator:Get({}))
+  end)
+  device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY*2, function(d)
+    device:send(SensorMultilevel:Get({}))
+  end)
+  device.thread:call_with_delay(constants.DEFAULT_GET_STATUS_DELAY*3, function(d)
+    device:send(Configuration:BulkGetV2({parameter_offset = 1, number_of_parameters = 6}) )
+  end)
+end
+
+--- Handle Device Added Event
+---
+--- @param driver st.zwave.Driver
+--- @param device st.zwave.Device
+local function device_added(driver, device)
+  device:send(BarrierOperator:Get({}))
+  -- Reset contact sensor battery level... This should be pollable on the GDO side.
+  device:emit_event_for_endpoint(CONTACTSENSOR_ENDPOINT_NUMBER,
+    capabilities.battery.battery(CONTACTSENSOR_BATTERY_LEVEL_NORMAL))
+  -- Reset contact sensor fields
+  device:emit_event_for_endpoint(CONTACTSENSOR_ENDPOINT_NUMBER,
+    capabilities.tamperAlert.tamper.clear())
+  device:emit_event_for_endpoint(CONTACTSENSOR_ENDPOINT_NUMBER,
+    capabilities.contactSensor.contact.closed())
+  -- Init barrier door state
+  device:emit_event_for_endpoint(GDO_ENDPOINT_NUMBER,
+    capabilities.doorControl.door.closed())
+end
+
+--- Configuration Report Handler
+---
+--- @param device st.zwave.Device
+local function configure_device_with_updated_config(device)
+  local updated_params = {}
+  updated_params[GDO_CONFIG_PARAM_NO_UNATTENDED_WAIT] = {parameter = device.preferences.closeWaitPeriodSec}
+  updated_params[GDO_CONFIG_PARAM_NO_ACTIVATION_TIME] = {parameter = device.preferences.activationTimeMS}
+  updated_params[GDO_CONFIG_PARAM_NO_OPEN_TIMEOUT] = {parameter = device.preferences.doorOpenTimeoutSec}
+  updated_params[GDO_CONFIG_PARAM_NO_CLOSE_TIMEOUT] = {parameter = device.preferences.doorCloseTimeoutSec}
+  updated_params[GDO_CONFIG_PARAM_NO_SHAKE_SENSE] = {parameter = device.preferences.shakeSensitivity}
+  updated_params[GDO_CONFIG_PARAM_NO_APP_RETRY] = {parameter = device.preferences.applicationLevelRetries}
```

Comment:
not necessary to change, but I think a more elegant way to do this would have been if you had defined a map at the top like 

```
local CONFIG_PARAMS = {
  closeWaitPeriodSec = 1,
  activationTimeMS = 2,
  ...
}
```

then this function could have been 

```
for param, value in pairs(device.preferences) do
  updated_params[CONFIG_PARAMS[param]] = value
end
```

Reply:
I agree, I made the change.